# GPT-2 Exploration

This notebook explores a pretrained GPT-2 model and compares
its tokenization, architecture, attention mechanisms, and
text generation with our Tiny GPT poem generator.

In [2]:
!pip install transformers torch

In [3]:
import torch
from transformers import GPT2Tokenizer, GPT2LMHeadModel

In [4]:
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
print(tokenizer)

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

GPT2Tokenizer(name_or_path='gpt2', vocab_size=50257, model_max_length=1024, padding_side='right', truncation_side='right', special_tokens={'bos_token': '<|endoftext|>', 'eos_token': '<|endoftext|>', 'unk_token': '<|endoftext|>'}, added_tokens_decoder={
	50256: AddedToken("<|endoftext|>", rstrip=False, lstrip=False, single_word=False, normalized=True, special=True),
})


## 2. Exploring GPT-2 Tokenization

GPT-2 does not process raw text directly. It first converts text into
tokens, and then converts those tokens into numerical token IDs.

We will examine:
- The original text
- The tokens created by GPT-2
- The numerical token IDs
- How the IDs can be converted back into text

In [5]:
prompt = "The moon shines over the hills"

# Convert text into GPT-2 tokens
tokens = tokenizer.tokenize(prompt)

# Convert text into numerical token IDs
token_ids = tokenizer.encode(prompt)

# Convert token IDs back into text
decoded_text = tokenizer.decode(token_ids)

print("Original text:")
print(prompt)

print("\nTokens:")
print(tokens)

print("\nToken IDs:")
print(token_ids)

print("\nDecoded text:")
print(decoded_text)

Original text:
The moon shines over the hills

Tokens:
['The', 'Ġmoon', 'Ġshines', 'Ġover', 'Ġthe', 'Ġhills']

Token IDs:
[464, 8824, 32481, 625, 262, 18639]

Decoded text:
The moon shines over the hills


## 2.5 Understanding GPT-2's Byte-Level BPE Tokenization

GPT-2 uses byte-level Byte Pair Encoding (BPE).

Instead of requiring every possible word to exist as a separate vocabulary
entry, GPT-2 can represent words using smaller subword pieces.

Common words may be represented by one token, while uncommon or longer
words may be split into multiple tokens.

This allows GPT-2 to handle a very large variety of words while using a
fixed vocabulary of 50,257 tokens.

In [7]:
examples = [
    "moon",
    "beautiful",
    "unbelievable",
    "Kigali",
    "transformer",
    "internationalization"
]

for text in examples:
    tokens = tokenizer.tokenize(text)

    print(f"Text: {text}")
    print(f"Tokens: {tokens}")
    print(f"Number of tokens: {len(tokens)}")
    print()

Text: moon
Tokens: ['moon']
Number of tokens: 1

Text: beautiful
Tokens: ['beaut', 'iful']
Number of tokens: 2

Text: unbelievable
Tokens: ['un', 'bel', 'iev', 'able']
Number of tokens: 4

Text: Kigali
Tokens: ['K', 'ig', 'ali']
Number of tokens: 3

Text: transformer
Tokens: ['trans', 'former']
Number of tokens: 2

Text: internationalization
Tokens: ['international', 'ization']
Number of tokens: 2



## 4. Loading the Pretrained GPT-2 Model

GPT-2 is a pretrained Transformer language model.

Unlike Tiny GPT, we are not training GPT-2 from scratch here.
We will open the pretrained model and inspect its architecture,
embeddings, attention heads, and text-generation behavior.

In [6]:
# Load the pretrained GPT-2 model
model = GPT2LMHeadModel.from_pretrained("gpt2")

# Put the model in evaluation mode
model.eval()

# Select GPU if available, otherwise use CPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Move the model to the selected device
model = model.to(device)

print("Device:", device)
print("Model loaded successfully!")

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors: reconstructing file:   0%|          |  0.00B /  548MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/148 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

Device: cuda
Model loaded successfully!


In [9]:


print("Original text:")
print(sample)

sample = "Love is"

# Tiny GPT tokenization: character-level
tiny_tokens = list(sample)

# GPT-2 tokenization: BPE
gpt2_tokens = tokenizer.tokenize(sample)

print("Original:", sample)

print("\nTiny GPT")
print("Tokens:", tiny_tokens)
print("Number of tokens:", len(tiny_tokens))

print("\nGPT-2")
print("Tokens:", gpt2_tokens)
print("Number of tokens:", len(gpt2_tokens))

Original text:
Love is
Original: Love is

Tiny GPT
Tokens: ['L', 'o', 'v', 'e', ' ', 'i', 's']
Number of tokens: 7

GPT-2
Tokens: ['Love', 'Ġis']
Number of tokens: 2


### 2.8 Tiny GPT vs GPT-2 Tokenization

Both models were given the same input: "Love is".

Tiny GPT uses character-level tokenization, producing 7 tokens:
['L', 'o', 'v', 'e', ' ', 'i', 's']

GPT-2 uses Byte-Level BPE tokenization, producing only 2 tokens:
['Love', 'Ġis']

This shows that Tiny GPT represents text character by character,
while GPT-2 can represent common words or subword units as larger
tokens. Therefore, GPT-2 generally requires fewer tokens to represent
the same natural-language text.

In [10]:
!pwd

/content
